perceval/utils/parameter.py

```
@@ -212,16 +218,197 @@ def max(self, m: bool):
+            return Expression(f"({self.name}^{other.name})", {self, other})
+        elif isinstance(other, (int, float)):
+            return Expression(f"({self.name}^{other})", {self})
+        raise TypeError("Unsupported parameter operation.")
+    
+    def __neg__(self):
+        # Ensure using __neg__ twice in a row returns the original parameter
+        if self._original is not None:
+            return self._original
+        
+        if isinstance(self, Expression):
+            expr = Expression(f"(-{self.name})", self._params)
+        else:
+            expr = Expression(f"(-{self.name})", {self})
+        expr._original = self
+        return expr
 
 class Expression(Parameter):
-    def __init__(self, expression: str):
+    """
+    This class allows arithmetic manipulation of the Parameter class. 
+    A logical string is passed and the parameters with a corresponding name are created.
+    Alternatively, one can specify the pre-defined parameters.
+    
+    :param name: string specifying equation, acts as name of Expression parameter.
+    :param parameters: specifies the identities of existing parameters present in the expression name 
+    """
+    def __init__(self, name: str, parameters = None):
         try:
-            e = sp.S(expression)
-        except sp.SympifyError as err:
-            raise ValueError("%s is not an expression: %s", expression, str(err))
-        assert isinstance(e, sp.Expr), "%s is not an expression" % expression
-        super().__init__("_%d" % Parameter._id, e, is_expression=True)
+            e = sp.S(name)
+            self.name = f"({e})"
+        except Exception as err:
+            raise ValueError("%s is not an expression: %s", name, str(err))
+        assert isinstance(e, sp.Expr), "%s is not an expression" % name
+        
+        # Create set containing all parent parameters
+        self._params = set() if parameters is None else set(parameters)
+        self._create_missing_parameters()
```

I'm not sure if that's better to autogenerate missing parameters, or to force the user to create them all (I'm feeling more confident to force the user to create all parameters, that way if they mess up their expression, they'll know it asap - at Expression instantiation - rather than at evaluation. It's better to fail fast).

On top of that, I tried this code:
```python
X, Y = P("X"), P("Y")
expr = E("2 * cos(X) + Y", {X, Y})
print(expr.parameters)

X.set_value(np.pi/4)
Y.set_value(3)
print(float(expr))
print(2*np.cos(np.pi/4) + 3)
```
which fails because of the `self._create_missing_parameters()` call which detects a "cos" parameter and creates it. This "cos" parameter shadows the Sympy cosinus function.
It works well with "2 * cos(X) + Y" as long as I comment out `self._create_missing_parameters()`.

To conclude this long comment, for the 2 reasons above, I'd suggest removing the feature to autogenerate the missing parameters.

---

Hi Eric. Thanks for the helpful feedback!

I think you're right regarding the auto-generation of missing parameters. I can take it out.